zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{traits::LiquiditySharesManager, BalanceOf, Config, Error};
+use alloc::{vec, vec::Vec};
+use core::marker::PhantomData;
+use frame_support::{
+    ensure,
+    pallet_prelude::RuntimeDebugNoBound,
+    storage::{bounded_btree_map::BoundedBTreeMap, bounded_vec::BoundedVec},
+    traits::Get,
+    PalletError,
+};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{
+    traits::{AtLeast32BitUnsigned, CheckedSub, Zero},
+    DispatchError, DispatchResult,
+};
+use zeitgeist_primitives::math::{
+    checked_ops_res::{CheckedAddRes, CheckedMulRes, CheckedSubRes},
+    fixed::FixedMulDiv,
+};
+
+/// Gets the maximum number of nodes allowed in the liquidity tree as a function of its depth.
+/// Saturates at `u32::MAX`.
+///
+/// # Generics
+///
+/// - `D`: A getter for the depth of the tree.
+pub struct LiquidityTreeMaxNodes<D>(PhantomData<D>);
+
+impl<D> Get<u32> for LiquidityTreeMaxNodes<D>
+where
+    D: Get<u32>,
+{
+    fn get() -> u32 {
+        2u32.saturating_pow(D::get() + 1).saturating_sub(1)
```

Comment:
come on @maltekliemann :D just use the possibility of writing into the debug assert instead of commenting it after the debug_assert:

```rust
debug_assert!(D::get() < 16, "D seems to be configured too high.");
```

Reply:
0826bbe92f7961082eb8599eca413110706d348f